### PR TITLE
ramping consolidation

### PR DESCRIPTION
**Pull request by daschuer**
This is the final PR about ramping and fixes 
https://bugs.launchpad.net/mixxx/+bug/1435170
https://bugs.launchpad.net/mixxx/+bug/1433457
https://bugs.launchpad.net/mixxx/+bug/1442831

It is on top of https://github.com/mixxxdj/mixxx/pull/546.

It removes the artificial samples and the fixed length of start and stop ramps, since we do not have buffers size independent ramps elsewhere in Mixxx.   

All rate-depended gain changes now at a common place in EnginePregain.
IMHO this is the best solution for a good maintenance.

### CODE REVIEW

**Comment by ywwg** [comment]
These new changes have made the PR too complicated for me to review, sorry.

**Comment by ywwg** [comment]
What I think we should do is create a way to test eniginebuffer.cpp similar to the way I added a test for the SoundTouch crash.  We create a fake readahead manager that generates a sine wave, and then use that instead of the actual RAMAN.  (Or we can create a sinewave .WAV file and load that).  Then there can be a bunch of tests for all these various edge cases of enginebuffer processing.  In the test we set up the conditions, call ProcessBuffer(), and then compare the actual buffer against a known-good "golden data" set.  enginebuffer should be deterministic so the tests should keep passing (maybe we can have a small per-sample error rate).  If the tests ever fail, we can compare the actual output with the golden output and see if there's a real problem or just a slight change that's ok.  Then we'll have much better confidence that our changes are working, and not breaking anything.

**Comment by daschuer** [comment]
> These new changes have made the PR too complicated for me to review, sorry.

That is why I kept https://github.com/mixxxdj/mixxx/pull/546, so lets merge those first.

**Comment by daschuer** [comment]
Testing this is reasonable, I think we can it make petty reliable if we use a square wave test signal that matches the buffer size. 

But before I start spending my time for this i need to be sure that it is worth the work.  We need to clarify that it is OK do do all the rate related ramping in EnginePregain. Please also check if it is OK to remove the (not working) buffer size independent ramp when coming from or going to pause. It is usually not played to the audience anyway. I also would like to merge the library sorting without tests.
See my reasons at https://github.com/mixxxdj/mixxx/pull/465.

**Comment by ywwg** [comment]
I don't think a square wave is a good idea because it will already have very sharp characteristics, and that's what we want these tests to find.  A triangle might be better, but I've had the best luck with a sine.

What I'd like to do is do the tests first so that we have a known-good baseline before we continue to move things around.  If we discover specific bugs as part of creating the tests we can fix them. Regardless of how we end up arranging the engine, the tests will be worth the work because we will be able to protect against regressions.

**Comment by daschuer** [comment]
@ywwg: Of cause we can move around the code when we discover issues, but we will not discover design issues. because the test will only test the current design. So it is important to answer the questions above before starting the work. Are my proposals OK for now?

**Comment by ywwg** [comment]
exactly, I want to test the current design and then see how it changes with your proposals.  Tests first, then changes.

**Comment by daschuer** [comment]
Test for the new design will fail with the current design.
I will not write test for the current design. Would you mind to answer my questions to continue here?

**Comment by ywwg** [comment]
Yeah the idea is then I can see how the new design is different from the current design.   The actual tests won't need to change, but the golden data might, and that's ok.  These changes are just getting to a point where I'm not confident in my ability to review them and catch any bugs there might be.  

If you write tests for your new design, we can apply them to the current design and just see what has changed.  Like, if the rate is X and the next buffer is paused, what does it look like?  It should fade out, no matter what design is being used.  So it's not like work done to test the current design will be thrown out.

**Comment by daschuer** [comment]
> It should fade out, no matter what design is being used.

This PR and the current version fades out different. We have also an open discussion where and if we need a fade though zero when scratching. It is not a matter of a test to decide that. 
I would also just merge https://github.com/mixxxdj/mixxx/pull/465 as it is, to save my time for this here. I need an explicit answerer to all three questions.

**Comment by ywwg** [comment]
> This PR and the current version fades out different.

yes and I want the tests so I can see what the difference is.  I'm not sure what the problem is with creating tests to see how the code behaves, and how the new code changes that behavior.

**Comment by daschuer** [comment]
A automatic test cannot achieve this. You can only test it by listening or watch a recorded sine wave in Audacity. (I have already done it and I am pleased, except of loop ramping which will be issued later) Once we have decided how it should be, we are able to write a test that verifies the result and keeps the behavior for later changes. 

The different is that in the master version, we have a constant ramp length of 3,3 ms when pressing play and pause. The ramp is overlayed with the vinyl ramp of on buffer size. This effectively limit the ramp to the buffers size below 3.3 ms if the buffer is shorter. Instead of a desired linear ramp it is somehow quadratic.

I have removed the play pause ramp and let only do the vinyl ramp its job. I have also removed the artificial samples after pause. 

So you probably now whant to listen to this PR and to the mater. If you have a conclusion, I will verify it with a test. 

Please answer my other questions as well. Thank you.

**Comment by ywwg** [comment]
> watch a recorded sine wave in Audacity

That's the idea, I want to create a test that processes a sine wave.  Then it won't be necessary to open audacity and check it manually.  That will be a huge win for future engine development.

It's going to take some time for me to review this, and these are all low-importance bugs in a high-risk part of the code, so it's more important to be safe.  Similarly the multi-sort header PR is a feature request, so my first priority is to make sure it doesn't break anything.  "Just merge it now" is a recipe for regressions and bugs, which has happened many many times in the recent past.  Sorry I don't have specific answers for you right now.

**Comment by daschuer** [comment]
> ... Then it won't be necessary to open audacity and check it manually. That will be a huge win for future engine development.

I fully agree, and I will do it. But that will only work if we decide about how should the good sample look like. So we need a decision first. 

I have already split the PR it into two to make review more easy, how about decide about the crossing zero fade first. If you agreed with the place in EnginePregain, I can add a test to https://github.com/mixxxdj/mixxx/pull/546 and we can merge it.

The multi-sort header PR was already reviewed and tested from three persons. 
A unit-test will be a lot of work and will be obsolete after the library refactoring. 
So we are sure by a high grade that it does not brake anything. 
Lets merge it and spend our time to more important topics. Is it OK for you? 





It's going to take some time for me to review this, and these are all low-importance bugs in a high-risk part of the code, so it's more important to be safe. Similarly the multi-sort header PR is a feature request, so my first priority is to make sure it doesn't break anything. "Just merge it now" is a recipe for regressions and bugs, which has happened many many times in the recent past. Sorry I don't have specific answers for you right now.

File: src/basetrackplayer.cpp [hunk 1]
```diff
@@ -166,7 +166,11 @@ void BaseTrackPlayerImpl::slotLoadTrack(TrackPointer track, bool bPlay) {
         disconnect(m_pLoadedTrack.data(), 0, this, 0);
         disconnect(m_pLoadedTrack.data(), 0, m_pKey, 0);
 
-        m_pReplayGain->slotSet(0);
+        // Do not reset ReplayGain here, because the track might be still
+        // playing and the last buffer will be processed.
+        //m_pReplayGain->slotSet(0);
```
Comment (ywwg) [inline src/basetrackplayer.cpp]: remove commented code

File: src/engine/enginebuffer.cpp [hunk 2]
```diff
@@ -1034,83 +1031,30 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
...
-        // the track and trying to jog forward so this uniquely blocks that
-        // situation.
-        if (fabs(rate) > kSmallRate && !bCurBufferPaused &&
-            m_iRampState != ENGINE_RAMP_UP && m_fRampValue < 1.0) {
-            m_iRampState = ENGINE_RAMP_UP;
-        }
-    }
+        // release the pauselock
+        m_pause.unlock();
+    } else { // if (!bTrackLoading && m_pause.tryLock()) {
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: remove commented code

File: src/engine/enginebuffer.cpp [hunk 3]
```diff
@@ -1034,83 +1031,30 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
...
-                pOutput[i] = pOutput[i] * m_fRampValue;
-                pOutput[i+1] = pOutput[i+1] * m_fRampValue;
-            }
+        // Here the old track was playing and loading the new track is in
+        // progress so we have not the chance to collect real samples for
+        // fade out.
+        // We apply a rectangular Gain change here which may click.
+        // It is in the responds of the user to not play this to
+        // the audience.
+        //if (m_speed_old) {
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: remove commented code.

I'm not sure I understand the sentence: "It is in the responds of the user to not play this to
     // the audience."

"It is the responsibility of the user not to play this to the audience"?  Why don't we fade the last sample to zero?  Or fade from the last sample into the new buffer?  It's not a "correct" fade but it works ok

File: src/engine/enginebuffer.cpp (hunk 2 above)
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: This is a useful comment since the corresponding "if" is 300 lines above.

File: src/engine/enginebuffer.cpp (hunk 3 above)
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: We can't predict when it happens. So we are not able to collect old samples. New samples are also not in place and we can't predict when they will be in place.  

This case happens only if it is allowed to load a track into a playing track. If one does this, a click from breaking the last track is somehow natural and he should know that such sound should not be played to the master (audiance). 
Simply pause the track before. 

A click free solution requires more refectory, which is out of the scope of this PR.  
Fading the last sample to zero clicks too.

Comment (daschuer) [inline src/engine/enginebuffer.cpp]: I will update the comment.

**Comment by daschuer** [comment]
Notes addressed. 
Note: Recent changes of https://github.com/mixxxdj/mixxx/pull/546 are not included to avoid criss cross merge.

**Comment by daschuer** [comment]
This is again ready for review.

**Comment by daschuer** [comment]
This is almost reviewed and rotting for more than year. 
The critical questions are already solved in PR https://github.com/mixxxdj/mixxx/pull/546 and already merged. 
So I would lieke to merge this soon. 
Any objections?

**Comment by daschuer** [comment]
Thank you for review so far. If here are remaining issues, we can fix them in master.
